edgelet/edgelet-http-workload/src/server/cert/identity.rs

```diff
@@ -1,19 +1,656 @@
+                                match cert_req.expiration() {
+                                    None => Ok(max_duration),
+                                    Some(exp) => compute_validity(exp, max_duration),
+                                }.map(|expiration| expiration)
+                                .map_err(Error::from)
+                            }).and_then(move |expiration| {
+                                hsm.destroy_certificate(alias.clone())
+                                    .map_err(Error::from)?;
+                                let sans = vec![module_uri];
+                                let props = CertificateProperties::new(
+                                    ensure_range!(expiration, 0, max_duration) as u64,
+                                    ensure_not_empty!(cn),
+                                    CertificateType::Client,
+                                    alias,
+                                ).with_san_entries(sans);
+                                hsm.create_certificate(&props)
+                                    .map_err(Error::from)
+                                    .and_then(|cert| {
+                                        let cert = cert_to_response(&cert)?;
+                                        let body = serde_json::to_string(&cert)?;
+                                        Response::builder()
+                                            .status(StatusCode::CREATED)
+                                            .header(CONTENT_TYPE, "application/json")
+                                            .header(CONTENT_LENGTH, body.len().to_string().as_str())
+                                            .body(body.into())
+                                            .map_err(From::from)
+                                    })
+                            }).unwrap_or_else(|e| e.into_response())
+                    }).map_err(Error::from)
+                    .or_else(|e| future::ok(e.into_response()));
+
+                future::Either::A(result)
+            }).unwrap_or_else(|e| future::Either::B(future::ok(e.into_response())));
+
+        Box::new(response)
+    }
+}
+
+fn cert_to_response<T: Certificate>(cert: &T) -> Result<CertificateResponse> {
+    let cert_buffer = cert.pem()?;
+    let expiration = cert.get_valid_to()?;
+
+    let private_key = match cert.get_private_key()? {
+        Some(PrivateKey::Ref(ref_)) => PrivateKeyResponse::new("ref".to_string()).with_ref(ref_),
+        Some(PrivateKey::Key(KeyBytes::Pem(buffer))) => PrivateKeyResponse::new("key".to_string())
+            .with_bytes(String::from_utf8_lossy(buffer.as_ref()).to_string()),
+        None => Err(ErrorKind::BadPrivateKey)?,
+    };
+
+    Ok(CertificateResponse::new(
+        private_key,
+        String::from_utf8_lossy(cert_buffer.as_ref()).to_string(),
+        expiration.to_rfc3339(),
+    ))
+}
+
+fn compute_validity(expiration: &str, max_duration_sec: i64) -> Result<i64> {
```

Comment:
This function also seems to be basically the same as the corresponding version from `./server.rs` (except for the additional `ensure_not_empty!` check. Re-use?